Make tests pass against current rgeo

dschweisguth opened the pull request:
Tests failed for me due only to incorrect setup, so document correct setup.
En passant, clean up Gemfile/gemspect and .gitignore.
Fixes https://github.com/rgeo/activerecord-mysql2spatial-adapter/issues/24

dschweisguth commented:
@januszm I closed my previous PR and opened this one. Please take a look!

januszm commented on History.rdoc:
```diff
@@ -1,3 +1,7 @@
+=== 0.5.1 / 2017-06-23
+
+* Constrain rgeo to compatible versions (< 0.5)
```
I think that we should say `< 0.6` here , you use 0.6 in Appraisal

```
gemfiles/rgeo_0.6.0.gemfile
```

dschweisguth replied:
Good catch. 0.6.0 is current rgeo so there is no longer a constraint on rgeo at all and nothing to document.

However, I realized that this PR exposes the mysql constraint to users. See the updated title and description.

januszm commented:
Looks good @dschweisguth , just let me do one thing before we merge this. I want to update Readme about tests for v 0.5.0 and release proper 0.5.0 version to Rubygems (current one is marked as `norelease`) .

dschweisguth commented:
I updated History.rdoc and removed the stray semicolon from the gemspec.

@januszm after you update the README I'll update it here to document the use of the appraisal gem. (I don't think we'll need to maintain rgeo 0.4.0 compatibility, but I think it will be useful to add other appraisals in the future, and other rgeo gems use the appraisal gem.)

januszm commented:
@dschweisguth I just released final 0.5.0 version to Rubygems (updated README and fixed mysql2 dependency) https://rubygems.org/gems/activerecord-mysql2spatial-adapter/versions/0.5.0

Please rebase your code on the current origin/master branch and let's continue with adding support for AR 4.2 and 5+. As I mentioned before, I think that it is perfectly fine to drop support for older versions (people can always lock gem version at 0.5.0).

I also started a code cleanup work on branch `code_cleanup`. We don't need to support old Ruby versions and I'd like to change the code to use "new" Ruby hash syntax, cleanup whitespace and newline characters etc.

dschweisguth commented:
@januszm this change is rebased and ready to go. Since the mysql2 version restriction is now in both Gemfile and gemspec in 0.5.0, this change doesn't have any user-visible effects, so I removed the version bump.

januszm commented:
@dschweisguth   Please add `gemfiles/*.lock` to `.gitignore` and remove those .lock files from Git.

I also left comment in the code/files changed.

januszm commented on test/tc_basic.rb:
```diff
@@ -114,7 +114,7 @@ def test_set_and_get_point_from_wkt
               assert_nil(obj_.latlon)
               obj_.latlon = 'SRID=1000;POINT(1 2)'
-              assert_equal(@factory.point(1, 2), obj_.latlon)
+              assert_equal(RGeo::Cartesian.preferred_factory(srid: 1000).point(1, 2), obj_.latlon)
```
Why we need to change this?

dschweisguth replied:
`@factory` has an SRID other than 1000. rgeo 0.4 incorrectly considered two points equal if their other properties were correct but their factories were different. rgeo 0.6 correctly includes factories when determining equality. This change makes the factory on each side of the comparison equal so that the points are equal.

januszm replied:
@dschweisguth if this change is about supporting rgeo 0.4 then I'd skip it. It seems that all our tests pass with rgeo 0.6.

Projects that depend on rgeo-activerecord branch 1.0 (like ours) can use rgeo 0.6 most probably, as rgeo-activerecord 1.x has a dependency on all rgeo 0.x versions:

```
  spec.add_dependency 'rgeo', '~> 0.3'
```

So, I'd vote for removing this change from the test suite.

januszm replied:
It's a bit confusing, I actually reverted your commit 3295269ebd372728948845121147b8a66e4c76f5 and ran the adapter test suite against both rgeo `0.4.0` and `0.6.0`. In both cases tests passed.

dschweisguth replied:
That IS confusing. On current master, or on this branch when I revert tc_basic.rb to current master, the three tests that I had changed fail:

```
[dave hladik:~/data/projects/activerecord-mysql2spatial-adapter] be rake test
Loaded testcase test/tc_basic.rb
Loaded testcase test/tc_spatial_queries.rb
Run options: --seed 35120

# Running:

.F.....F.F....

Finished in 3.045653s, 4.5967 runs/s, 8.5368 assertions/s.

  1) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_save_and_load_point_from_wkt [test/tc_basic.rb:141]:
No visible difference in the RGeo::Cartesian::PointImpl#inspect output.
You should look at the implementation of #== on RGeo::Cartesian::PointImpl or its members.
#<RGeo::Cartesian::PointImpl:0xXXXXXX "POINT (1.0 2.0)">

  2) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_readme_example [test/tc_basic.rb:165]:
Expected: true
  Actual: false

  3) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_set_and_get_point_from_wkt [test/tc_basic.rb:117]:
No visible difference in the RGeo::Cartesian::PointImpl#inspect output.
You should look at the implementation of #== on RGeo::Cartesian::PointImpl or its members.
#<RGeo::Cartesian::PointImpl:0xXXXXXX "POINT (1.0 2.0)">

14 runs, 26 assertions, 3 failures, 0 errors, 0 skips
```

When I constrain rgeo to 0.4.* these tests pass for me.

dschweisguth replied:
@januszm here's my Gemfile.lock:

```
GIT
  remote: https://github.com/rgeo/rgeo-activerecord.git
  revision: 0d3f7975b476135b7baf8b1831b4fa2528500616
  branch: 1.0
  specs:
    rgeo-activerecord (1.2.0)
      activerecord (~> 4.0)
      rgeo (~> 0.3)

PATH
  remote: .
  specs:
    activerecord-mysql2spatial-adapter (0.5.0)
      activerecord (>= 4.0, < 4.2)
      mysql2 (>= 0.2.13, < 0.4.0)
      rgeo-activerecord (~> 1.0)

GEM
  remote: https://rubygems.org/
  specs:
    activemodel (4.1.16)
      activesupport (= 4.1.16)
      builder (~> 3.1)
    activerecord (4.1.16)
      activemodel (= 4.1.16)
      activesupport (= 4.1.16)
      arel (~> 5.0.0)
    activesupport (4.1.16)
      i18n (~> 0.6, >= 0.6.9)
      json (~> 1.7, >= 1.7.7)
      minitest (~> 5.1)
      thread_safe (~> 0.1)
      tzinfo (~> 1.1)
    appraisal (2.2.0)
      bundler
      rake
      thor (>= 0.14.0)
    arel (5.0.1.20140414130214)
    builder (3.2.3)
    coderay (1.1.1)
    i18n (0.8.4)
    json (1.8.6)
    method_source (0.8.2)
    minitest (5.10.2)
    mysql2 (0.3.21)
    pry (0.10.4)
      coderay (~> 1.1.0)
      method_source (~> 0.8.1)
      slop (~> 3.4)
    rake (12.0.0)
    rdoc (5.1.0)
    rgeo (0.6.0)
    slop (3.6.0)
    thor (0.19.4)
    thread_safe (0.3.6)
    tzinfo (1.2.3)
      thread_safe (~> 0.1)

PLATFORMS
  ruby

DEPENDENCIES
  activerecord-mysql2spatial-adapter!
  appraisal
  pry
  rake (>= 0.9.2)
  rdoc (>= 3.12)
  rgeo-activerecord!

BUNDLED WITH
   1.15.1
```

Also, I installed GEOS to see if it would fix any of the failures; it did not.

januszm replied:
weird, I copied your Gemfile.lock and all tests pass:

```bash
bundle exec rake
rm -rf doc
rm -rf pkg
rm -rf tmp
Loaded testcase test/tc_basic.rb
Loaded testcase test/tc_spatial_queries.rb
Run options: --seed 14699

# Running:

..............

Finished in 0.219661s, 63.7345 runs/s, 127.4690 assertions/s.

14 runs, 28 assertions, 0 failures, 0 errors, 0 skips
```

dschweisguth replied:
@januszm I'm at a loss as to where to look. I'm on macOS 10.12.5 with current XCode and mysql 5.6.30. I'm using Ruby 2.3.4 but saw the same results on a couple of other versions. I delete and recreate gemsets and `bundle clean --force` often, so there are no gems or gem versions in my gemset other than those in that Gemfile.lock.

Just to be clear, I meant that I see the failures on my branch (whose gemspec results in that Gemfile.lock) when I roll back my change to tc_basic.rb. But you probably already understood that, because why else would I have made the changes?

dschweisguth replied:
@januszm Looking closer at the tests, the from_wkt tests fail for me because, in the comparison at lines 117 and 141, the two factories have different SRIDs. rgeo 0.4 considered the factories equal despite the different SRIDs; rgeo 0.6 considers them different (which makes sense to me). In rgeo 0.6 the second comparison in RGeo::ImplHelper::BasicPointMethods:67 fails. What's going on there in your environment?

januszm replied:
I'm on Mac OS X 10.11, Ruby 2.3.3, MySQL 5.7

please take a look at this, debugger inside test ~ line 117

```ruby
> @factory
=> #<RGeo::Geos::CAPIFactory:0x3feb86068d9c srid=3785 bufres=1 flags=8>
> RGeo::Cartesian.preferred_factory(srid: 1000)
=> #<RGeo::Geos::CAPIFactory:0x3feb85d28100 srid=1000 bufres=1 flags=8>
> assert_equal(@factory.point(1, 2), obj_.latlon)
=> true
> assert_equal(1000, obj_.latlon.srid)
=> true
> RGeo::VERSION
=> "0.6.0"
```

You can see that factories differ in srid attribute (3785 vs 1000) but assertions return true

dschweisguth replied:
@januszm `RGeo::Cartesian.preferred_factory` returns a `RGeo::Cartesian::Factory` for me, which includes SRID in equality. This is because `RGeo::Geos.supported?` returns false for me. `RGeo::Geos::FFI_SUPPORTED` and `RGeo::Geos::CAPI_SUPPORTED` are both false for me; which is true for you, and why? Do you have GEOS installed? (As I said above I didn't, and installing it didn't change this behavior. Not sure why not, actually, but let's see what your situation is first.)

januszm replied:
@dschweisguth yes I had geos installed, version 3.5. After uninstalling geos those tests fail:

```bash
  1) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_readme_example [test/tc_basic.rb:165]:
Expected: true
  Actual: false

  2) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_save_and_load_point_from_wkt [test/tc_basic.rb:141]:
No visible difference in the RGeo::Cartesian::PointImpl#inspect output.
You should look at the implementation of #== on RGeo::Cartesian::PointImpl or its members.
#<RGeo::Cartesian::PointImpl:0xXXXXXX "POINT (1.0 2.0)">

  3) Failure:
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_set_and_get_point_from_wkt [test/tc_basic.rb:117]:
No visible difference in the RGeo::Cartesian::PointImpl#inspect output.
You should look at the implementation of #== on RGeo::Cartesian::PointImpl or its members.
#<RGeo::Cartesian::PointImpl:0xXXXXXX "POINT (1.0 2.0)">
```

I installed it again (this time version 3.6) and tests pass again.

Now, installation of `geos` is 'highly' recommended. For older versions of rgeo gem it wasn't even possible to install it without geos present.

To sum up, I see two distinct topics in this PR:
1. gemspec, Gemfile and gitignore cleanup, which could be done in just 2 commits (1 for gems, 1 for gitignore)
2. dealing with rgeo and geos - honestly, I'd skip that and just focus on support for ActiveRecord 4.2+

You already spent some time on this and I think that it'd be nice if your commits make their way to master. Could you please remove commits related to rgeo from this branch and squash it to 2 commits. Then we could release 0.5.1 version of the gem.

januszm commented on Gemfile:
```diff
@@ -31,17 +31,11 @@
-end
+# This dependency is here because gemspecs can't use git sources.
+# It overrides the less specific dependency in gemspec.
+gem 'rgeo-activerecord', git: 'https://github.com/rgeo/rgeo-activerecord.git', branch: '1.0'
```
To be honest, it seems that we can remove it now. `rgeo-activerecord` v1.2 is available at rubygems.org since August 29, so after I introduced this dependency, and that's just what we need now.

dschweisguth replied:
It seems we still need the constraint in the Gemfile. Without it, one test fails:

```
RGeo::ActiveRecord::Mysql2SpatialAdapter::Tests::TestBasic#test_create_geometry_with_index:
ArgumentError: struct size differs
```

Looks like your last contribution to that branch didn't make it in to the last rgeo-activerecord 1.* release (1.2.0).

januszm replied:
@dschweisguth OK, now it's released :), https://rubygems.org/gems/rgeo-activerecord/versions/1.3.0

dschweisguth replied:
I updated the gemspec to require that version.

dschweisguth commented:
@januszm .lock files cleaned up, other comments responded to in comments.

dschweisguth commented:
@januszm here's what I have left after removing the now-unnecessary test changes. I also removed the use of appraisal since tests pass against the current resolved versions of all dependencies.

Please note that I'll be away soon for about a week, during which I won't be able to respond to this sort of thing.

dschweisguth commented:
@januszm I'm back. How does this look?

januszm commented:
@dschweisguth merged and published to RubyGems!
